### PR TITLE
Add support for deleting resources in reverse order by stage.

### PR DESCRIPTION
**Pull request by hankfreund**
### Change description

- Add finalizers to expanded resources and process deletion requests
- Track expanded objects by stage
- When deleting, iterate over the stages in reverse order and delete all objects. Once those objects no longer exist, proceed to the next stage
- Remove owner reference injection on expanded resources

### CODE REVIEW

**Review by barney-s — COMMENTED** [review]
It would be curious to see how this code will behave with Updates to composition.

Like an user changes composition to add or remove stages.
Or even re-order stages !!

**Comment by barney-s** [comment]
Foreground with ownership may may not solve ordering deletion issue: 

https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#ownership-and-finalizers

File: experiments/compositions/composition/tests/scenario/scenario.go [hunk 1]
```diff
@@ -154,11 +154,11 @@ func NewKCCSample(t *testing.T, sample Sample, dependentSamples []Sample) *Scena
 }
 
 func (s *Scenario) Cleanup() {
+	defer cluster.ReleaseCluster(s.T, s.cluster)
+	defer s.GatherLogs()
```
[the remark this answers is not in the document]
Comment (hankfreund) [reply experiments/compositions/composition/tests/scenario/scenario.go:158]: If I'm not mistaken, defers are executed in LIFO order, so gather will be called before release.

File: experiments/compositions/composition/internal/controller/expander_reconciler.go [hunk 2]
```diff
@@ -645,6 +701,8 @@ func (r *ExpanderReconciler) enqueueAllFromGVK(ctx context.Context, _ client.Obj
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ExpanderReconciler) SetupWithManager(mgr ctrl.Manager, cr *unstructured.Unstructured) error {
+	r.appliedObjects = map[types.NamespacedName][]*appliedStage{}
```
[the remark this answers is not in the document]
Comment (hankfreund) [reply experiments/compositions/composition/internal/controller/expander_reconciler.go:704]: I have an idea, we can discuss offline.

File: experiments/compositions/composition/tests/testcases/simple_test.go [hunk 3]
```diff
@@ -455,3 +436,56 @@ func TestUpdateCompositionModifyStage(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("failed reading configmap: %v", err)
+	}
+	cms[1] = tmpCm
+	controllerutil.RemoveFinalizer(cms[1], "test/delay")
+	if err := s.C.Update(context.Background(), cms[1]); err != nil {
+		t.Fatalf("failed removing finalizer from configmap: %v", err)
+	}
+
+	// Verify the remaining objects have been deleted
```
[the remark this answers is not in the document]
Comment (hankfreund) [reply experiments/compositions/composition/tests/testcases/simple_test.go:476]: Line 479 will handle that check -- should I modify it to match this?

**Pull request closed without merging**